pytential/symbolic/pde/system_utils.py

```diff
@@ -0,0 +1,466 @@
+        # TargetMultipliers
+        new_int_gs = convert_target_multiplier_to_source(expr)
+        # Convert IntGs with different kernels to expressions containing
+        # IntGs with base_kernel or its derivatives
+        return sum(convert_int_g_to_base(new_int_g,
+            self.base_kernel) for new_int_g in new_int_gs)
+
+
+def convert_target_deriv_to_source(int_g):
+    """Converts AxisTargetDerivatives to AxisSourceDerivative instances
+    from an IntG. If there are outer TargetPointMultiplier transformations
+    they are preserved.
+    """
+    knl = int_g.target_kernel
+    source_kernels = list(int_g.source_kernels)
+    coeff = 1
+    multipliers = []
+    while isinstance(knl, TargetPointMultiplier):
+        multipliers.append(knl.axis)
+        knl = knl.inner_kernel
+
+    while isinstance(knl, AxisTargetDerivative):
+        coeff *= -1
+        source_kernels = [AxisSourceDerivative(knl.axis, source_knl) for
+                source_knl in source_kernels]
+        knl = knl.inner_kernel
+
+    # TargetPointMultiplier has to be the outermost kernel
+    # If it is the inner kernel, return early
+    if isinstance(knl, TargetPointMultiplier):
+        return int_g
+
+    for axis in reversed(multipliers):
+        knl = TargetPointMultiplier(axis, knl)
+
+    new_densities = tuple(density*coeff for density in int_g.densities)
+    return int_g.copy(target_kernel=knl,
+                      densities=new_densities,
+                      source_kernels=tuple(source_kernels))
+
+
+def _get_kernel_expression(expr, kernel_arguments):
+    from pymbolic.mapper.substitutor import substitute
+    from sumpy.symbolic import PymbolicToSympyMapperWithSymbols
+
+    pymbolic_expr = substitute(expr, kernel_arguments)
+
+    res = PymbolicToSympyMapperWithSymbols()(pymbolic_expr)
+    return res
+
+
+def _monom_to_expr(monom, variables):
+    prod = 1
+    for i, nrepeats in enumerate(monom):
+        for _ in range(nrepeats):
+            prod *= variables[i]
+    return prod
+
+
+def convert_target_multiplier_to_source(int_g):
+    """Convert an IntG with TargetMultiplier to a sum of IntGs without
+    TargetMultiplier and only source dependent transformations
+    """
+    import sympy
+    import sumpy.symbolic as sym
+    from sumpy.symbolic import SympyToPymbolicMapper
+    conv = SympyToPymbolicMapper()
+
+    knl = int_g.target_kernel
+    # we use a symbol for d = (x - y)
+    ds = sympy.symbols(f"d0:{knl.dim}")
+    sources = sympy.symbols(f"y0:{knl.dim}")
+    # instead of just x, we use x = (d + y)
+    targets = [d + source for d, source in zip(ds, sources)]
+    orig_expr = sympy.Function("f")(*ds)  # pylint: disable=not-callable
+    expr = orig_expr
+    found = False
+    while isinstance(knl, KernelWrapper):
+        if isinstance(knl, TargetPointMultiplier):
+            expr = targets[knl.axis] * expr
+            found = True
+        elif isinstance(knl, AxisTargetDerivative):
```

Comment:
Above, everything gets normalized to source derivatives, and then those don't get handled...?

Reply:
Fixed in eec08e26